File: DS/Linkedlist/singly_linked_list/PYTHON/Recursion-problems/print_list.py
class Node:
    def __init__(self, num):
        self.data = num
        self.link = None


class LinkedList:

    def __init__(self):
        self.head = None
        
    # insert node
    def append(self, num):
        temp = Node(num)
        if self.head is None:
            self.head = temp
            return

        temp1 = self.head
        while temp1.link is not None:
            temp1 = temp1.link

        temp1.link = temp

# print list
def print_list(pointer):
    if pointer is None:
        return
    print(pointer.data, end=" ")
    print_list(pointer.link)

# print list in reverse order
def print_reverse(pointer):
    if pointer is None:
        return
    print_reverse(pointer.link)
    print(pointer.data, end=" ")


if __name__ == '__main__':

    l_list = LinkedList()
    
    l_list.append(1)
    l_list.append(2)
    l_list.append(3)
    l_list.append(4)
    l_list.append(5)

    print_list(l_list.head)
    print_reverse(l_list.head)

File: DS/Linkedlist/singly_linked_list/JAVA/Recursion_problems/print_list.java
public class Linkedlist {

    static class Node
    {
        int data;
        Node link;

        Node(int num){
            data = num;
            link = null;
        }
    }
    Node head;

    //insert a node at the end of linked list
    public void append(int num){
        Node temp = new Node(num);
        if(head == null){
            head = temp;
            return;
        }
        Node temp1 = head;
        while (temp1.link!=null){
            temp1 = temp1.link;
        }

        temp1.link = temp;
    }

    //print list
    public void print(Node pointer){
        if(pointer==null) return;

        System.out.print(pointer.data + " ");
        print(pointer.link);
    }

    //print list in reverse order
    public void reversePrint(Node pointer){
        if(pointer==null) return;

        reversePrint(pointer.link);
        System.out.print(pointer.data + " ");
    }

    //main
    public static void main(String[] args){
        linkedlist list = new linkedlist();
        list.head = null;
        
        list.append(1);
        list.append(2);
        list.append(3);
        list.append(4);
        list.append(5);
        
        list.print(list.head);
        System.out.print("\n");
        list.reversePrint(list.head);
    }
}
